### PR TITLE
ARROW-15635: [C++] Support nested extension-id-registry

### PR DESCRIPTION
**Pull request by rtpsw**
See [this post](https://issues.apache.org/jira/browse/ARROW-16211?focusedCommentId=17539044&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17539044).

### CODE REVIEW

**Comment by rtpsw** [comment]
cc @icexelloss

**Comment by rtpsw** [comment]
If this makes sense to other people, I could also try to make a similar separate solution for the function registry.

**Comment by vibhatha** [comment]
@rtpsw Is the PR associated with the correct JIRA?

File: cpp/src/arrow/engine/substrait/extension_set.h [hunk 1]
```diff
@@ -224,7 +240,7 @@ class ARROW_ENGINE_EXPORT ExtensionSet {
   std::size_t num_functions() const { return functions_.size(); }
 
  private:
-  ExtensionIdRegistry* registry_;
+  const ExtensionIdRegistry* registry_;
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/extension_set.h:243]: Is there any particular reason to change this to a `const`?  I think it's fine but it seems unrelated to the change so I wanted to make sure I wasn't missing anything.

**Review by westonpace — CHANGES_REQUESTED** [review]
I think the concept is a good idea.  I think @sanjibansg is close to having a PR for ARROW-15582 which changes the extension id registry a bit.  However, these two PRs are addressing orthogonal concerns so it should be pretty straightforward to rebase whichever one ends up being second.

What would you envision the lifetime of the custom extension id registry being?  Based on some of the comments it sounds like you might be thinking per-plan.

For a plan-specific "embedded UDF" (i.e. a UDF that has been pickled and serialized as part of the plan) I though the idea was that the embedded UDF would have some special way of being inserted into an expression (admittedly, the protobuf is lacking for this part at the moment) and so it wouldn't need to be a part of the extension set.

For UDFs that are not embedded in the plan itself it seems the user would probably want to just register those once so I think the lifetime of this nested set would still be process-scoped.

The proposed change works fine for process-scoped but I want to make sure I understand the intended usage.

Lastly, we should have some unit tests in place before we merge this.

**Comment by rtpsw** [comment]
> @rtpsw Is the PR associated with the correct JIRA?

Yeah, this is by mistake. I'll move it soon.

**Comment by rtpsw** [comment]
> What would you envision the lifetime of the custom extension id registry being? Based on some of the comments it sounds like you might be thinking per-plan.

Either or both per-process and per-plan. The scope can be controlled by the user (from Arrow or PyArrow):
- For per-process scope, the user makes `per_process_registry = nested_extension_id_registry(default_extension_id_registry())` and keeps it throughout the lifetime of the process.
- For per-plan scope, the user recreates `per_plan_registry = nested_extension_id_registry(default_extension_id_registry())` for each plan, 
- For both per-process and per-plan scope, the user makes `per_process_registry = nested_extension_id_registry(default_extension_id_registry())` and recreates `per_plan_registry = nested_extension_id_registry(per_process_registry)` (note the double-nesting) for each plan.
 
> For a plan-specific "embedded UDF" (i.e. a UDF that has been pickled and serialized as part of the plan) I though the idea was that the embedded UDF would have some special way of being inserted into an expression (admittedly, the protobuf is lacking for this part at the moment) and so it wouldn't need to be a part of the extension set.

Locally, I was able to get this to work for UDFs restricted to element-wise flat-scalar-valued (not analytic or reduction, not struct, not table or dataset).

> For UDFs that are not embedded in the plan itself it seems the user would probably want to just register those once so I think the lifetime of this nested set would still be process-scoped.

My local solution involves an Ibis/Substrait/Arrow workflow, where all UDFs are registered once and get serialized into each Substrait plan they are used in.

> The proposed change works fine for process-scoped but I want to make sure I understand the intended usage.

As explained above, I think it works more broadly.

> Lastly, we should have some unit tests in place before we merge this.

Agreed. I'll move on to add them, now that this approach seems to be acceptable.

**Comment by rtpsw** [comment]
This PR is replaced by https://github.com/apache/arrow/pull/13232
